src/symbol/collision_index.js

```diff
@@ -0,0 +1,346 @@
+
+        const projectionCache = {};
+        const fontScale = fontSize / 24;
+        const lineOffsetX = symbol.lineOffsetX * fontSize;
+        const lineOffsetY = symbol.lineOffsetY * fontSize;
+
+        const tileUnitAnchorPoint = new Point(symbol.anchorX, symbol.anchorY);
+        const labelPlaneAnchorPoint = projection.project(tileUnitAnchorPoint, labelPlaneMatrix).point;
+        const firstAndLastGlyph = projection.placeFirstAndLastGlyph(
+            fontScale,
+            glyphOffsetArray,
+            lineOffsetX,
+            lineOffsetY,
+            /*flip*/ false,
+            labelPlaneAnchorPoint,
+            tileUnitAnchorPoint,
+            symbol,
+            lineVertexArray,
+            labelPlaneMatrix,
+            projectionCache,
+            /*return tile distance*/ true);
+
+        if (projectedAnchor.perspectiveRatio < .2 || !firstAndLastGlyph) {
+            // This label either doesn't fit on its line or is way off screen: mark it unused.
+            for (let k = 0; k < collisionCircles.length; k += 5) {
+                collisionCircles[k + 4] = true;
+            }
+            return [];
+        }
+
+        let collisionDetected = false;
+
+        const tileToViewport = projectedAnchor.perspectiveRatio * pixelsToTileUnits;
+        const tileToViewportScaled = tileToViewport * scale;
+
+        let firstTileDistance = 0, lastTileDistance = 0;
+        if (firstAndLastGlyph) {
+            firstTileDistance = this.approximateTileDistance(firstAndLastGlyph.first.tileDistance, firstAndLastGlyph.first.angle, tileToViewport, projectedAnchor.cameraDistance, pitchWithMap);
+            lastTileDistance = this.approximateTileDistance(firstAndLastGlyph.last.tileDistance, firstAndLastGlyph.last.angle, tileToViewport, projectedAnchor.cameraDistance, pitchWithMap);
+        }
+
+        for (let k = 0; k < collisionCircles.length; k += 5) {
+            const boxDistanceToAnchor = collisionCircles[k + 3];
+            const tileUnitRadius = collisionCircles[k + 2];
+            if ((boxDistanceToAnchor < -firstTileDistance) ||
+                (boxDistanceToAnchor - tileUnitRadius > lastTileDistance)) {
+                // Don't need to use this circle because the label doesn't extend this far
+                collisionCircles[k + 4] = true;
+                continue;
+            }
+
+            const projectedPoint = this.projectPoint(posMatrix, collisionCircles[k], collisionCircles[k + 1]);
+            const x = projectedPoint.x;
+            const y = projectedPoint.y;
+
+            const radius = tileUnitRadius / tileToViewportScaled;
+
+            if (placedCollisionCircles.length) {
+                const dx = x - placedCollisionCircles[placedCollisionCircles.length - 4];
+                const dy = y - placedCollisionCircles[placedCollisionCircles.length - 3];
+                if (radius * radius * 2 > dx * dx + dy * dy) {
```

Comment:
This is `if (radius * sqrt(2) > distance(prev, current)` right? Is the `sqrt(2)` just an experimentally derived value?

Reply:
Yeah, I ended up at this number with just a few rough experiments looking at how the circle density turned out. We could probably come up with a more theoretically optimal number.

Reply:
I updated this section with comments and hopefully more self-explanatory variable names, and I revisited the √2 value experimentally. Using a value of 1 is significantly slower, and by the time I upped the value to √3, I started to see missed collision detections (usually with orthogonal streets that "passed through" each other).